server/src/main/java/org/tctalent/server/service/db/impl/LinkPreviewServiceImpl.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright (c) 2024 Talent Beyond Boundaries.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see https://www.gnu.org/licenses/.
+ */
+
+package org.tctalent.server.service.db.impl;
+
+import static java.lang.Double.parseDouble;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.jsoup.Connection;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.select.Elements;
+import org.springframework.lang.Nullable;
+import org.springframework.stereotype.Service;
+import org.tctalent.server.exception.EntityReferencedException;
+import org.tctalent.server.exception.InvalidRequestException;
+import org.tctalent.server.logging.LogBuilder;
+import org.tctalent.server.model.db.LinkPreview;
+import org.tctalent.server.repository.db.LinkPreviewRepository;
+import org.tctalent.server.service.db.LinkPreviewService;
+
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class LinkPreviewServiceImpl implements LinkPreviewService {
```

Comment:
A lot of the logic in here is hard to follow without a broad understanding how previews are generated - a link to a doc like https://andrejgajdos.com/how-to-create-a-link-preview/ would be helpful to any developer coming into this code cold and trying to add or fix something.

Reply:
@camerojo I actually already linked to that article and gave a little doc overview in the service interface: should I add some to the implementation, too?

Absolutely agree on the entities — I'll merge this and make a separate issue to add some doc.